sls_api/endpoints/tools/files.py

```diff
@@ -443,6 +448,216 @@ def get_file_tree(project, file_path=None):
+
+    # Resolve the real, absolute paths
+    base_dir = os.path.realpath(config["file_root"])
+    full_path = os.path.realpath(full_path)
+
+    # Verify that full_path is within base_dir, i.e. file_root specified
+    # in config
+    if os.path.commonpath([base_dir, full_path]) != base_dir:
+        return create_error_response("Error: invalid file path.", 400)
+
+    # Check if the file exists
+    try:
+        if not os.path.isfile(full_path):
+            return create_error_response("Error: the requested file was not found on the server.", 404)
+    except Exception:
+        logger.exception(f"Error accessing file at {full_path}")
+        return create_error_response(f"Error accessing file at {file_path}", 500)
+
+    # Check that the file has a .xml extension
+    if not full_path.endswith(".xml"):
+        return create_error_response("Error: the file path must point to a file with a .xml extension.", 400)
+
+    # Check file size so we don't parse overly large XML files
+    max_file_size = 5 * 1024 * 1024  # 5 MB
+    if os.path.getsize(full_path) > max_file_size:
+        return create_error_response("Error: file size exceeds the maximum allowed limit (5 MB).", 400)
+
+    # Process the XML file
+    metadata, error_message, status_code = extract_publication_metadata_from_tei_xml(full_path)
+    if error_message:
+        return create_error_response(error_message, status_code=status_code)
+    else:
+        return create_success_response("Metadata retrieved from XML file.", data=metadata)
+
+
+def extract_publication_metadata_from_tei_xml(file_path: str) -> Tuple[Optional[Dict[str, Any]], Optional[str], Optional[int]]:
+    """
+    Extracts publication metadata (document title, date of origin and main
+    language) from a TEI XML file located at the given file path.
+
+    Args:
+
+        file_path (str): The absolute path to the TEI XML file.
+
+    Returns:
+
+    - A tuple containing:
+        - metadata (dict or None): A dictionary with the extracted
+          metadata:
+            - "name" (str or None): The title extracted from the XML.
+            - "original_publication_date" (str or None): The date of
+              origin in "YYYY", "YYYY-MM", or "YYYY-MM-DD" format.
+            - "language" (str or None): The language code.
+            - "genre" (str or None): The genre of the text.
+            Returns `None` if an error occurred.
+        - error_message (str or None): An error message if an error
+          occurred; otherwise `None`.
+        - status_code (int or None): The HTTP status code corresponding to
+        the result (e.g., 200 on success, 404 if file not found).
+
+    Examples:
+
+        >>> metadata, error_message, status_code = extract_publication_metadata_from_tei_xml('/path/to/file.xml')
+    """
+    try:
+        # Parse the XML file and extract relevant metadata from it
+        with open(file_path, "r", encoding="utf-8-sig") as xml_file:
```

Comment:
Are the XML files in question *always* utf-8-sig encoded? What happens if they are plain utf-8?

Reply:
The 'utf-8-sig' codec can read UTF-8-encoded files written with and without the starting BOM signature and will remove it if present, so since we are reading file contents here, this works for both UTF-8 and UTF-8-BOM encoded files. Our XML guidelines state that the files should be saved without BOM, but you never know ...